src/client/linter.ts

```diff
@@ -0,0 +1,134 @@
+	Position,
+	Range,
+	DiagnosticSeverity,
+	Diagnostic,
+	languages,
+	Uri
+} from "vscode";
+import { tmpdir, networkInterfaces } from "os";
+import { sep } from "path";
+import { arrayInclude, trimBoth, getWorkspaceFolder } from "./utils";
+import { execV } from "./exec";
+import { resolve, relative, dirname } from "path";
+import { readdirSync } from "fs";
+
+const outDir = `${tmpdir()}${sep}vscode_vlang${sep}`;
+const SEV_ERR = DiagnosticSeverity.Error;
+const SEV_WRN = DiagnosticSeverity.Warning;
+
+export interface ErrorInfo {
+	file: string;
+	line: number;
+	column: number;
+	message: string;
+	stderr: string;
+}
+
+interface MoreInfo {
+	for: number;
+	content: string;
+}
+
+export const collection = languages.createDiagnosticCollection("V");
+
+export function lint(document: TextDocument): boolean {
+	const workspaceFolder = getWorkspaceFolder(document.uri);
+	// Don't lint files that are not in the workspace
+	if (!workspaceFolder) return true;
+
+	const cwd = workspaceFolder.uri.fsPath;
+	const foldername = dirname(document.fileName);
+	const relativeFoldername = relative(cwd, foldername);
+	const relativeFilename = relative(cwd, document.fileName);
+	const fileCount = readdirSync(foldername).filter(f => f.endsWith(".v")).length;
+
+	let target = foldername === cwd ? "." : relativeFoldername;
+	target = fileCount === 1 ? relativeFilename : target;
+
+	let status = true;
+
+	execV(`-o ${outDir}lint.c ${target}`, (err, stdout, stderr) => {
+		if (err || stderr.trim().length > 1) {
+			const output = stderr || stdout;
+			const isWarning = output.substring(0, 7) === "warning";
+
+			if (!isWarning) {
+				/* ERROR */
+				const { file, line, column, message } = parseError(output);
+				const fileuri = Uri.file(resolve(cwd, file));
+				const start = new Position(line - 1, column);
+				const end = new Position(line - 1, column + 1);
+				const range = new Range(start, end);
+				const diagnostic = new Diagnostic(range, message, SEV_ERR);
+				diagnostic.source = "V";
+				collection.set(fileuri, [diagnostic]);
+			} else {
+				/* WARNING */
+				collection.clear();
+				const warnings = parseWarning(output);
+				warnings.forEach(warning => {
+					const { file, line, column, message } = warning;
+					const fileuri = Uri.file(resolve(cwd, file));
+					const start = new Position(line - 1, column);
+					const end = new Position(line - 1, column + 1);
+					const range = new Range(start, end);
+					const diagnostic = new Diagnostic(range, message, SEV_WRN);
+					diagnostic.source = "V";
+					collection.set(fileuri, [...collection.get(fileuri), diagnostic]);
+				});
+			}
+			return (status = false);
+		} else {
+			collection.delete(document.uri);
+		}
+	});
+	return status;
+}
+
+function parseWarning(stderr: string): Array<ErrorInfo> {
+	stderr = trimBoth(stderr);
+	const lines = stderr.split("\n");
+	const warnings: Array<ErrorInfo> = [];
+	const moreInfos: Array<MoreInfo> = [];
+
+	for (let i = 0; i < lines.length; i++) {
```

Comment:
It would make more sense to use `for(let ln of lines)` than a traditional C-style `for` loop.